Enhancing User Experience by Generating Security Audit Reports in PDF, DOC, or TXT Formats

Currently, the output of the NPM package is displayed in the VSCode terminal after running the required commands. However, for end users and developers (who downloads the NPM package), it would be more convenient if the security audit report could be generated as a PDF, DOC, or TXT file, allowing for easier review and avoiding the need to scroll through the terminal output.

Sample terminal report: 
[audit report.pdf](https://github.com/user-attachments/files/17230062/audit.report.pdf)

**Make sure the end user/developer (who downloads the NPM package) is able to smoothly run the NPM package after these changes.**

when i run src/NodeJsSecurify.js file I get this output in the terminal 
`******************************************************************************************
****************************** Node-Js-Securify STARTED **********************************
******************************************************************************************

Searching for .js files in (root directory) : /home/eleensmathew/NodejsSecurify/src

File names in .gitignore files not getting parsed: 

, Vulnerability, dist, NodeJsSecurify.js, node_modules

File path name of .js & .jsx files getting parsed: 

/home/eleensmathew/NodejsSecurify/src/GenerateReport.js
==> No callback hell detected. Code looks good!


==> Code NOT vulnerable to Brute force Attack


==> RegExp object detected with ReDoS-prone patterns at line 18


==> Code having proper input validation


==> No dangerous function detected


==> NO Insecure Authentication detected


==> Code NOT vulnerable to Brute force Attack


/home/eleensmathew/NodejsSecurify/src/GenerateReport.ts
Error: Line 2: Unexpected token =
    at ErrorHandler.constructError (/home/eleensmathew/NodejsSecurify/node_modules/esprima/dist/esprima.js:5012:22)
    at ErrorHandler.createError (/home/eleensmathew/NodejsSecurify/node_modules/esprima/dist/esprima.js:5028:27)
    at Parser.throwError (/home/eleensmathew/NodejsSecurify/node_modules/esprima/dist/esprima.js:1936:34)
    at Parser.parseImportDeclaration (/home/eleensmathew/NodejsSecurify/node_modules/esprima/dist/esprima.js:4837:23)
    at Parser.parseStatementListItem (/home/eleensmathew/NodejsSecurify/node_modules/esprima/dist/esprima.js:3373:39)
    at Parser.parseScript (/home/eleensmathew/NodejsSecurify/node_modules/esprima/dist/esprima.js:4723:29)
    at parse (/home/eleensmathew/NodejsSecurify/node_modules/esprima/dist/esprima.js:122:61)
    at Object.parseScript (/home/eleensmathew/NodejsSecurify/node_modules/esprima/dist/esprima.js:145:13)
    at Function.<anonymous> (/home/eleensmathew/NodejsSecurify/src/NodeJsSecurify.js:177:101)
    at step (/home/eleensmathew/NodejsSecurify/src/NodeJsSecurify.js:49:23) {
  index: 44,
  lineNumber: 2,
  description: 'Unexpected token ='
}
/home/eleensmathew/NodejsSecurify/src/NodeJsSecurify.ts
Error: Line 36: Unexpected token :
    at ErrorHandler.constructError (/home/eleensmathew/NodejsSecurify/node_modules/esprima/dist/esprima.js:5012:22)
    at ErrorHandler.createError (/home/eleensmathew/NodejsSecurify/node_modules/esprima/dist/esprima.js:5028:27)
    at Parser.unexpectedTokenError (/home/eleensmathew/NodejsSecurify/node_modules/esprima/dist/esprima.js:1985:39)
    at Parser.throwUnexpectedToken (/home/eleensmathew/NodejsSecurify/node_modules/esprima/dist/esprima.js:1995:21)
    at Parser.expect (/home/eleensmathew/NodejsSecurify/node_modules/esprima/dist/esprima.js:2168:19)
    at Parser.parseFormalParameters (/home/eleensmathew/NodejsSecurify/node_modules/esprima/dist/esprima.js:4274:23)
    at Parser.parseFunctionDeclaration (/home/eleensmathew/NodejsSecurify/node_modules/esprima/dist/esprima.js:4337:38)
    at Parser.parseStatementListItem (/home/eleensmathew/NodejsSecurify/node_modules/esprima/dist/esprima.js:3379:39)
    at Parser.parseFunctionSourceElements (/home/eleensmathew/NodejsSecurify/node_modules/esprima/dist/esprima.js:4195:29)
    at Parser.isolateCoverGrammar (/home/eleensmathew/NodejsSecurify/node_modules/esprima/dist/esprima.js:2269:37) {
  index: 1407,
  lineNumber: 36,
  description: 'Unexpected token :'
}
Parsing for vulnerable npm pacakage:
Error executing npm audit: Command failed: npm audit --json
npm notice Beginning October 4, 2021, all connections to the npm registry - including for package installation - must use TLS 1.2 or higher. You are currently using plaintext http to connect. Please visit the GitHub blog for more information: https://github.blog/2021-08-23-npm-registry-deprecating-tls-1-0-tls-1-1/
npm notice Beginning October 4, 2021, all connections to the npm registry - including for package installation - must use TLS 1.2 or higher. You are currently using plaintext http to connect. Please visit the GitHub blog for more information: https://github.blog/2021-08-23-npm-registry-deprecating-tls-1-0-tls-1-1/
npm WARN audit 426 Upgrade Required - POST http://registry.npmjs.org/-/npm/v1/security/audits/quick
npm ERR! audit endpoint returned an error` all the errors not printed using console.log i am not outputing to the file. If you want me to include the messages printed in the terminal using console.error please let me know. Also if you can give me a test file to test the code with that would be better

Updated pdf 
[NodeJsSecurifyReport.pdf](https://github.com/user-attachments/files/17275920/NodeJsSecurifyReport.pdf)

```
******************************************************************************************
****************************** Node-Js-Securify STARTED **********************************
******************************************************************************************

Searching for .js files in (root directory) : /home/eleensmathew/NodejsSecurify/TestFolder
``
File names in .gitignore files not getting parsed: 

, Vulnerability, dist, NodeJsSecurify.js, node_modules

File path name of .js & .jsx files getting parsed: 

/home/eleensmathew/NodejsSecurify/TestFolder/BruteForceAttack/BruteForceNotVulnerable-1.js
==> No callback hell detected. Code looks good!


==> Code NOT vulnerable to Brute force Attack


==> NO RegExp object detected with ReDoS-prone patterns


==> Code having proper input validation


==> No dangerous function detected


==> NO Insecure Authentication detected


==> Code NOT vulnerable to Brute force Attack


/home/eleensmathew/NodejsSecurify/TestFolder/BruteForceAttack/BruteForceVulnerable-1.js
==> No callback hell detected. Code looks good!


==> Code vulnerable to Brute force Attack in this file!!! 


==> NO RegExp object detected with ReDoS-prone patterns


==> Code having proper input validation


==> No dangerous function detected


==> NO Insecure Authentication detected


==> Code vulnerable to Brute force Attack in this file!!! 


/home/eleensmathew/NodejsSecurify/TestFolder/CallBackHell/CallBackHell-1.js
==> Callback hell detected! AT Line no. => 25 In file: CallBackHell-1.js
Make use of async/await or .then/.catch to avoid callback hell


==> Code NOT vulnerable to Brute force Attack


==> NO RegExp object detected with ReDoS-prone patterns


==> Code having proper input validation


==> Following dangerous function detected: 
> setTimeout detected at line: 1 Ensure: Remember to clear the interval when it's no longer needed: clearTimeout(timerId);
=====> Ignore above errors if you already made changes to your code <=====


==> NO Insecure Authentication detected


==> Code NOT vulnerable to Brute force Attack


/home/eleensmathew/NodejsSecurify/TestFolder/CallBackHell/CallBackHell-10.js
==> Callback hell detected! AT Line no. => 34 In file: CallBackHell-10.js
Make use of async/await or .then/.catch to avoid callback hell


==> Code NOT vulnerable to Brute force Attack


==> NO RegExp object detected with ReDoS-prone patterns


==> Code having proper input validation


==> Following dangerous function detected: 
> setTimeout detected at line: 1 Ensure: Remember to clear the interval when it's no longer needed: clearTimeout(timerId);
=====> Ignore above errors if you already made changes to your code <=====


==> NO Insecure Authentication detected


==> Code NOT vulnerable to Brute force Attack


/home/eleensmathew/NodejsSecurify/TestFolder/CallBackHell/CallBackHell-2.js
==> Callback hell detected! AT Line no. => 10 In file: CallBackHell-2.js
Make use of async/await or .then/.catch to avoid callback hell


==> Code vulnerable to Brute force Attack in this file!!! 


==> NO RegExp object detected with ReDoS-prone patterns


==> Code having proper input validation


==> No dangerous function detected


==> NO Insecure Authentication detected


==> Code vulnerable to Brute force Attack in this file!!! 


/home/eleensmathew/NodejsSecurify/TestFolder/CallBackHell/CallBackHell-4.js
==> Callback hell detected! AT Line no. => 6 In file: CallBackHell-4.js
Make use of async/await or .then/.catch to avoid callback hell


==> Code NOT vulnerable to Brute force Attack


==> NO RegExp object detected with ReDoS-prone patterns


==> Code having proper input validation


==> Following dangerous function detected: 
> setTimeout detected at line: 1 Ensure: Remember to clear the interval when it's no longer needed: clearTimeout(timerId);
=====> Ignore above errors if you already made changes to your code <=====


==> NO Insecure Authentication detected


==> Code NOT vulnerable to Brute force Attack


/home/eleensmathew/NodejsSecurify/TestFolder/CallBackHell/CallBackHell-5.js
==> Callback hell detected! AT Line no. => 8 In file: CallBackHell-5.js
Make use of async/await or .then/.catch to avoid callback hell
==> Callback hell detected! AT Line no. => 10 In file: CallBackHell-5.js
Make use of async/await or .then/.catch to avoid callback hell


==> Code vulnerable to Brute force Attack in this file!!! 


==> NO RegExp object detected with ReDoS-prone patterns


==> Code having proper input validation


==> No dangerous function detected


==> NO Insecure Authentication detected


==> Code NOT vulnerable to Brute force Attack


/home/eleensmathew/NodejsSecurify/TestFolder/CallBackHell/CallBackHell-7.js
==> Callback hell detected! AT Line no. => 4 In file: CallBackHell-7.js
Make use of async/await or .then/.catch to avoid callback hell


==> Code NOT vulnerable to Brute force Attack


==> NO RegExp object detected with ReDoS-prone patterns


==> Code having proper input validation


==> Following dangerous function detected: 
> getElementById detected at line: 1 Ensure: Regular expression to match only letters
=====> Ignore above errors if you already made changes to your code <=====


==> NO Insecure Authentication detected


==> Code NOT vulnerable to Brute force Attack


/home/eleensmathew/NodejsSecurify/TestFolder/CallBackHell/CallBackHell-9.js
==> Callback hell detected! AT Line no. => 4 In file: CallBackHell-9.js
Make use of async/await or .then/.catch to avoid callback hell


==> Code NOT vulnerable to Brute force Attack


==> NO RegExp object detected with ReDoS-prone patterns


==> Code having proper input validation


==> No dangerous function detected


==> Code vulnerable to Insecure Authentication in this file!!! 
Check for: Hardcoded Credentials, Weak Passwords, No Password Hashing, Insecure Session Management, Insecure Token Storage, Weak JWT Security


==> Code NOT vulnerable to Brute force Attack


/home/eleensmathew/NodejsSecurify/TestFolder/CallBackHell/CallBackHellNoError-3.js
==> No callback hell detected. Code looks good!


==> Code vulnerable to Brute force Attack in this file!!! 


==> NO RegExp object detected with ReDoS-prone patterns


==> Code having proper input validation


==> No dangerous function detected


==> NO Insecure Authentication detected


==> Code NOT vulnerable to Brute force Attack


/home/eleensmathew/NodejsSecurify/TestFolder/CallBackHell/CallBackHellNoError-6.js
==> No callback hell detected. Code looks good!


==> Code NOT vulnerable to Brute force Attack


==> NO RegExp object detected with ReDoS-prone patterns


==> Code having proper input validation


==> No dangerous function detected


==> NO Insecure Authentication detected


==> Code NOT vulnerable to Brute force Attack


/home/eleensmathew/NodejsSecurify/TestFolder/CallBackHell/CallBackHellNoError-8.js
==> No callback hell detected. Code looks good!


==> Code NOT vulnerable to Brute force Attack


==> NO RegExp object detected with ReDoS-prone patterns


==> Code having proper input validation


==> No dangerous function detected


==> NO Insecure Authentication detected


==> Code NOT vulnerable to Brute force Attack


/home/eleensmathew/NodejsSecurify/TestFolder/DetectDangerousFunctions/DangerousFunctions-1.js
==> No callback hell detected. Code looks good!


==> Code NOT vulnerable to Brute force Attack


==> RegExp object detected with ReDoS-prone patterns at line 9


==> Code having proper input validation


==> Following dangerous function detected: 
> setTimeout detected at line: 1 Ensure: Remember to clear the interval when it's no longer needed: clearTimeout(timerId);
> parsedInt detected at line: 1 Ensure: Always specify the radix (base) to avoid unexpected results
> parsedFloat detected at line: 1 Ensure: Always specify the radix (base) to avoid unexpected results
> getElementById detected at line: 1 Ensure: Regular expression to match only letters
> outerHTML detected at line: 1 Ensure: It does'nt accepts script tag, can cause XSS Attack
> innerHTML detected at line: 1 Ensure: It does'nt accepts script tag, can cause XSS Attack
> document.write detected at line: 1 Ensure: Avoid using it to prevent DOM XSS attack
=====> Ignore above errors if you already made changes to your code <=====


==> NO Insecure Authentication detected


==> Code NOT vulnerable to Brute force Attack


/home/eleensmathew/NodejsSecurify/TestFolder/InputValidation/InputValidationError.js
==> No callback hell detected. Code looks good!


==> Code NOT vulnerable to Brute force Attack


==> NO RegExp object detected with ReDoS-prone patterns


==> Code having improper input validation, vulnerable to DOS Attack:!!! 


==> No dangerous function detected


==> NO Insecure Authentication detected


==> Code NOT vulnerable to Brute force Attack


/home/eleensmathew/NodejsSecurify/TestFolder/InputValidation/NoInputValidationError-1.js
==> No callback hell detected. Code looks good!


==> Code NOT vulnerable to Brute force Attack


==> NO RegExp object detected with ReDoS-prone patterns


==> Code having proper input validation


==> No dangerous function detected


==> NO Insecure Authentication detected


==> Code NOT vulnerable to Brute force Attack


/home/eleensmathew/NodejsSecurify/TestFolder/InputValidation/NoInputValidationError-2.js
==> No callback hell detected. Code looks good!


==> Code NOT vulnerable to Brute force Attack


==> NO RegExp object detected with ReDoS-prone patterns


==> Code having proper input validation


==> No dangerous function detected


==> NO Insecure Authentication detected


==> Code NOT vulnerable to Brute force Attack


/home/eleensmathew/NodejsSecurify/TestFolder/InputValidation/NoInputValidationError-3.js
==> No callback hell detected. Code looks good!


==> Code NOT vulnerable to Brute force Attack


==> NO RegExp object detected with ReDoS-prone patterns


==> Code having proper input validation


==> No dangerous function detected


==> NO Insecure Authentication detected


==> Code NOT vulnerable to Brute force Attack


/home/eleensmathew/NodejsSecurify/TestFolder/InputValidation/NoInputValidationError-4.js
==> No callback hell detected. Code looks good!


==> Code NOT vulnerable to Brute force Attack


==> NO RegExp object detected with ReDoS-prone patterns


==> Code having proper input validation


==> No dangerous function detected


==> NO Insecure Authentication detected


==> Code NOT vulnerable to Brute force Attack


/home/eleensmathew/NodejsSecurify/TestFolder/InsecureAuthentication/InsecureAuthenticatio2.js
==> No callback hell detected. Code looks good!


==> Code NOT vulnerable to Brute force Attack


==> NO RegExp object detected with ReDoS-prone patterns


==> Code having proper input validation


==> Following dangerous function detected: 
> JSON.parse detected at line: 1 Ensure: Be cautious about the source of the JSON data, Can lead to  Client-side JSON injection
=====> Ignore above errors if you already made changes to your code <=====


==> NO Insecure Authentication detected


==> Code vulnerable to Brute force Attack in this file!!! 


/home/eleensmathew/NodejsSecurify/TestFolder/InsecureAuthentication/InsecureAuthentication1.js
==> No callback hell detected. Code looks good!


==> Code NOT vulnerable to Brute force Attack


==> NO RegExp object detected with ReDoS-prone patterns


==> Code having proper input validation


==> Following dangerous function detected: 
> JSON.parse detected at line: 1 Ensure: Be cautious about the source of the JSON data, Can lead to  Client-side JSON injection
=====> Ignore above errors if you already made changes to your code <=====


==> NO Insecure Authentication detected


==> Code vulnerable to Brute force Attack in this file!!! 


/home/eleensmathew/NodejsSecurify/TestFolder/InsecureAuthentication/InsecureAuthentication3.js
==> No callback hell detected. Code looks good!


==> Code NOT vulnerable to Brute force Attack


==> NO RegExp object detected with ReDoS-prone patterns


==> Code having proper input validation


==> No dangerous function detected


==> NO Insecure Authentication detected


==> Code NOT vulnerable to Brute force Attack


/home/eleensmathew/NodejsSecurify/TestFolder/InsecureAuthentication/InsecureAuthentication4.js
==> No callback hell detected. Code looks good!


==> Code NOT vulnerable to Brute force Attack


==> RegExp object detected with ReDoS-prone patterns at line 4


==> Code having proper input validation


==> No dangerous function detected


==> NO Insecure Authentication detected


==> Code NOT vulnerable to Brute force Attack


/home/eleensmathew/NodejsSecurify/TestFolder/InsecureAuthentication/InsecureAuthentication5.js
==> No callback hell detected. Code looks good!


==> Code NOT vulnerable to Brute force Attack


==> NO RegExp object detected with ReDoS-prone patterns


==> Code having proper input validation


==> No dangerous function detected


==> NO Insecure Authentication detected


==> Code NOT vulnerable to Brute force Attack


/home/eleensmathew/NodejsSecurify/TestFolder/InsecureAuthentication/InsecureAuthentication6.js
==> No callback hell detected. Code looks good!


==> Code NOT vulnerable to Brute force Attack


==> RegExp object detected with ReDoS-prone patterns at line 4


==> Code having proper input validation


==> No dangerous function detected


==> NO Insecure Authentication detected


==> Code NOT vulnerable to Brute force Attack


/home/eleensmathew/NodejsSecurify/TestFolder/InsecureAuthentication/InsecureAuthentication7.js
==> No callback hell detected. Code looks good!


==> Code vulnerable to Brute force Attack in this file!!! 


==> NO RegExp object detected with ReDoS-prone patterns


==> Code having proper input validation


==> No dangerous function detected


==> NO Insecure Authentication detected


==> Code NOT vulnerable to Brute force Attack


/home/eleensmathew/NodejsSecurify/TestFolder/InsecureSecurityHeader/InsecureSecurityHeader1.js
==> No callback hell detected. Code looks good!


==> Code NOT vulnerable to Brute force Attack


==> NO RegExp object detected with ReDoS-prone patterns


==> Code having improper input validation, vulnerable to DOS Attack:!!! 


==> No dangerous function detected


==> NO Insecure Authentication detected


==> Code vulnerable to Brute force Attack in this file!!! 


/home/eleensmathew/NodejsSecurify/TestFolder/InsecureSecurityHeader/InsecureSecurityHeader2.js
==> No callback hell detected. Code looks good!


==> Code NOT vulnerable to Brute force Attack


==> NO RegExp object detected with ReDoS-prone patterns


==> Code having improper input validation, vulnerable to DOS Attack:!!! 


==> No dangerous function detected


==> NO Insecure Authentication detected


==> Code NOT vulnerable to Brute force Attack


/home/eleensmathew/NodejsSecurify/TestFolder/InsecureSecurityHeader/InsecureSecurityHeader3.js
==> No callback hell detected. Code looks good!


==> Code NOT vulnerable to Brute force Attack


==> NO RegExp object detected with ReDoS-prone patterns


==> Code having improper input validation, vulnerable to DOS Attack:!!! 


==> No dangerous function detected


==> NO Insecure Authentication detected


==> Code vulnerable to Brute force Attack in this file!!! 


/home/eleensmathew/NodejsSecurify/TestFolder/InsecureSecurityHeader/InsecureSecurityHeader4.js
==> No callback hell detected. Code looks good!


==> Code NOT vulnerable to Brute force Attack


==> NO RegExp object detected with ReDoS-prone patterns


==> Code having improper input validation, vulnerable to DOS Attack:!!! 


==> No dangerous function detected


==> NO Insecure Authentication detected


==> Code NOT vulnerable to Brute force Attack


/home/eleensmathew/NodejsSecurify/TestFolder/InsecureSecurityHeader/InsecureSecurityHeader5.js
==> No callback hell detected. Code looks good!


==> Code NOT vulnerable to Brute force Attack


==> NO RegExp object detected with ReDoS-prone patterns


==> Code having improper input validation, vulnerable to DOS Attack:!!! 


==> No dangerous function detected


==> NO Insecure Authentication detected


==> Code NOT vulnerable to Brute force Attack


/home/eleensmathew/NodejsSecurify/TestFolder/RegexAttack/RegexAttack1.js
==> No callback hell detected. Code looks good!


==> Code NOT vulnerable to Brute force Attack


==> RegExp object detected with ReDoS-prone patterns at line 2


==> Code having proper input validation


==> No dangerous function detected


==> NO Insecure Authentication detected


==> Code NOT vulnerable to Brute force Attack


/home/eleensmathew/NodejsSecurify/TestFolder/RegexAttack/RegexAttack2.js
==> No callback hell detected. Code looks good!


==> Code vulnerable to Brute force Attack in this file!!! 


==> RegExp object detected with ReDoS-prone patterns at line 2


==> Code having proper input validation


==> No dangerous function detected


==> NO Insecure Authentication detected


==> Code NOT vulnerable to Brute force Attack


/home/eleensmathew/NodejsSecurify/TestFolder/RegexAttack/RegexAttack3.js
==> No callback hell detected. Code looks good!


==> Code NOT vulnerable to Brute force Attack


==> RegExp object detected with ReDoS-prone patterns at line 2


==> Code having proper input validation


==> No dangerous function detected


==> NO Insecure Authentication detected


==> Code vulnerable to Brute force Attack in this file!!! 


/home/eleensmathew/NodejsSecurify/TestFolder/RegexAttack/RegexAttack4.js
==> No callback hell detected. Code looks good!


==> Code NOT vulnerable to Brute force Attack


==> RegExp object detected with ReDoS-prone patterns at line 2


==> Code having proper input validation


==> No dangerous function detected


==> NO Insecure Authentication detected


==> Code NOT vulnerable to Brute force Attack


/home/eleensmathew/NodejsSecurify/TestFolder/RegexAttack/RegexAttack5.js
==> No callback hell detected. Code looks good!


==> Code NOT vulnerable to Brute force Attack


==> RegExp object detected with ReDoS-prone patterns at line 2


==> Code having proper input validation


==> No dangerous function detected


==> NO Insecure Authentication detected


==> Code NOT vulnerable to Brute force Attack


/home/eleensmathew/NodejsSecurify/TestFolder/RegexAttack/RegexAttack6.js
==> No callback hell detected. Code looks good!


==> Code NOT vulnerable to Brute force Attack


==> RegExp object detected with ReDoS-prone patterns at line 2


==> Code having proper input validation


==> No dangerous function detected


==> NO Insecure Authentication detected


==> Code NOT vulnerable to Brute force Attack


/home/eleensmathew/NodejsSecurify/TestFolder/RegexAttack/RegexAttack7.js
==> No callback hell detected. Code looks good!


==> Code NOT vulnerable to Brute force Attack


==> RegExp object detected with ReDoS-prone patterns at line 3


==> Code having proper input validation


==> No dangerous function detected


==> NO Insecure Authentication detected


==> Code NOT vulnerable to Brute force Attack


Parsing for vulnerable npm pacakage:
Error executing npm audit: Command failed: npm audit --json
npm notice Beginning October 4, 2021, all connections to the npm registry - including for package installation - must use TLS 1.2 or higher. You are currently using plaintext http to connect. Please visit the GitHub blog for more information: https://github.blog/2021-08-23-npm-registry-deprecating-tls-1-0-tls-1-1/
npm notice Beginning October 4, 2021, all connections to the npm registry - including for package installation - must use TLS 1.2 or higher. You are currently using plaintext http to connect. Please visit the GitHub blog for more information: https://github.blog/2021-08-23-npm-registry-deprecating-tls-1-0-tls-1-1/
npm WARN audit 426 Upgrade Required - POST http://registry.npmjs.org/-/npm/v1/security/audits/quick
npm ERR! audit endpoint returned an error
```

--- a/.gitignore
+++ b/.gitignore
@@ -3,3 +3,5 @@
 /publish.txt
 /img.jpg
 /dist
+/consoleOutput.log
+/NodeJsSecurifyReport.pdf

--- a/package.json
+++ b/package.json
@@ -34,6 +34,7 @@
     "@types/esprima": "^4.0.3",
     "@types/esquery": "^1.5.0",
     "@types/node": "^20.3.3",
+    "@types/pdfkit": "^0.13.5",
     "typescript": "^5.1.6"
   },
   "dependencies": {
@@ -43,6 +44,7 @@
     "esquery": "^1.5.0",
     "natural": "^6.5.0",
     "path": "^0.12.7",
+    "pdfkit": "^0.15.0",
     "randexp": "^0.5.3"
   },
   "files": [

--- a/src/GenerateReport.ts
+++ b/src/GenerateReport.ts
@@ -1,0 +1,54 @@
+import * as fs from 'fs';
+import PDFDocument = require('pdfkit');
+
+function generatePDFReport(): void {
+    const doc = new PDFDocument();
+    const output = fs.createWriteStream('NodeJsSecurifyReport.pdf');
+
+    doc.pipe(output);
+
+    doc.fontSize(15).text('NodeJsSecurify Report', { align: 'center' });
+    doc.moveDown();
+
+    const logContent: string = fs.readFileSync('consoleOutput.log', 'utf8');
+    function parseAnsiToPdf(doc: PDFKit.PDFDocument, content: string) {
+        const lines = content.split('\n');
+        lines.forEach(line => {
+            const regex = /\x1b\[(\d+)m/g;
+            let match;
+            let lastIndex = 0;
+            while ((match = regex.exec(line)) !== null) {
+                const colorCode = parseInt(match[1], 10);
+                const text = line.substring(lastIndex, match.index);
+                if (text) {
+                    doc.fontSize(10).text(text, { continued: true });
+                }
+                switch (colorCode) {
+                    case 30: doc.fillColor('black'); break;
+                    case 31: doc.fillColor('red'); break;
+                    case 32: doc.fillColor('green'); break;
+                    case 33: doc.fillColor('yellow'); break;
+                    case 34: doc.fillColor('blue'); break;
+                    case 35: doc.fillColor('magenta'); break;
+                    case 36: doc.fillColor('cyan'); break;
+                    case 37: doc.fillColor('white'); break;
+                    case 39: doc.fillColor('black'); break; // Reset to default color
+                    default: break;
+                }
+                lastIndex = regex.lastIndex;
+            }
+            const remainingText = line.substring(lastIndex);
+            if (remainingText) {
+                doc.text(remainingText);
+            }
+            doc.text('\n');
+        });
+    }
+
+    parseAnsiToPdf(doc, logContent);
+
+
+    doc.end();
+}
+
+generatePDFReport();

--- a/src/GenerateReport.ts
+++ b/src/GenerateReport.ts
@@ -10,7 +10,10 @@ function generatePDFReport(): void {
     doc.fontSize(15).text('NodeJsSecurify Report', { align: 'center' });
     doc.moveDown();
 
-    const logContent: string = fs.readFileSync('consoleOutput.log', 'utf8');
+    let logContent: string = fs.readFileSync('consoleOutput.log', 'utf8');
+    const lines: string[] = logContent.split('\n');
+    lines[2] = '****************************** Node-Js-Securify STARTED ***************************'
+    logContent = lines.join('\n');
     function parseAnsiToPdf(doc: PDFKit.PDFDocument, content: string) {
         const lines = content.split('\n');
         lines.forEach(line => {
@@ -27,7 +30,7 @@ function generatePDFReport(): void {
                     case 30: doc.fillColor('black'); break;
                     case 31: doc.fillColor('red'); break;
                     case 32: doc.fillColor('green'); break;
-                    case 33: doc.fillColor('yellow'); break;
+                    case 33: doc.fillColor('maroon'); break;
                     case 34: doc.fillColor('blue'); break;
                     case 35: doc.fillColor('magenta'); break;
                     case 36: doc.fillColor('cyan'); break;

--- a/src/NodeJsSecurify.ts
+++ b/src/NodeJsSecurify.ts
@@ -58,6 +58,7 @@ export class Log {
             };
 
             console.error = function () {
+                logFile.write(util.format.apply(null, Array.from(arguments)) + '\n');
                 logStdout.write(util.format.apply(null, Array.from(arguments)) + '\n');
             };
 

--- a/src/NodeJsSecurify.ts
+++ b/src/NodeJsSecurify.ts
@@ -19,6 +19,7 @@ import * as esprima from 'esprima';
 import * as fs from "fs";
 import * as path from 'path';
 import * as colors from 'colors';
+import * as util from 'util';
 import { detectBruteForce } from './Vulnerability/DetectBruteForceAttack'
 import { detectCallBackHell } from './Vulnerability/DetectCallBackHell';
 import { isRegexVulnerable } from './Vulnerability/DetectVulnerableRegex';
@@ -48,6 +49,18 @@ export class Log {
             return inputPath;
         }
         try {
+            const logFile = fs.createWriteStream('consoleOutput.log', { flags: 'w' });
+            const logStdout = process.stdout;
+
+            console.log = function () {
+                logFile.write(util.format.apply(null, Array.from(arguments)) + '\n');
+                logStdout.write(util.format.apply(null, Array.from(arguments)) + '\n');
+            };
+
+            console.error = function () {
+                logStdout.write(util.format.apply(null, Array.from(arguments)) + '\n');
+            };
+
             // testing command: node ./dist/NodeJsSecurify.js
             // comment this before publishing npm package (uncomment only for testing)
             // __dirname = "F:/NodeSecurify/TestFolder" // update this path depending on the path of TestFolder according to your system
@@ -77,6 +90,8 @@ export class Log {
             // parsing for vulnerable npm pacakage
             console.log("Parsing for vulnerable npm pacakage:".yellow);
             checkVulnerablePackages();
+
+            logFile.end();
         }
         catch (error: any) {
             console.log("Error parsing file".underline.red);

--- a/src/Vulnerability/DetectUnsafeNpmPackage.ts
+++ b/src/Vulnerability/DetectUnsafeNpmPackage.ts
@@ -1,9 +1,24 @@
 import { exec } from 'child_process';
+import * as fs from 'fs';
+import * as util from 'util';
 
 export function checkVulnerablePackages() {
   const auditCommand = 'npm audit --json';
   
   exec(auditCommand, (error, stdout, stderr) => {
+    const logFile = fs.createWriteStream('consoleOutput.log', { flags: 'a' });
+    const logStdout = process.stdout;
+
+    console.log = function () {
+        logFile.write(util.format.apply(null, Array.from(arguments)) + '\n');
+        logStdout.write(util.format.apply(null, Array.from(arguments)) + '\n');
+    };
+
+    console.error = function () {
+        logFile.write(util.format.apply(null, Array.from(arguments)) + '\n');
+        logStdout.write(util.format.apply(null, Array.from(arguments)) + '\n');
+    };
+    
     if (error) {
       console.error(`Error executing npm audit: ${error.message}`);
       return;
@@ -23,5 +38,6 @@ export function checkVulnerablePackages() {
     } else {
       console.log('No vulnerable packages found.'.green);
     }
+    logFile.end();
   });
 }

